### PR TITLE
Fix/missing timeout (#54)

### PR DESCRIPTION
**Pull request by siy**
* Enable screen/device lock timeout

### CODE REVIEW

**Comment by siy** [comment]
Testing exposed an [issue with USB](https://github.com/LedgerHQ/ledger-nanos-sdk/issues/91) when screen saver/lock is activated.

File: Cargo.toml [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 authors = ["siy"]
 edition = "2021"
 description = "Radix Babylon"
 
 [dependencies]
+ledger-sdk-sys = { git = "https://github.com/LedgerHQ/secure-sdk-rust" }
 nanos_sdk = { git = "https://github.com/LedgerHQ/ledger-nanos-sdk.git" }
-nanos_ui = { git = "https://github.com/radixdlt/ledger-nanos-ui-radix.git" }
+nanos_ui = { git = "https://github.com/LedgerHQ/ledger-nanos-ui.git" }
 include_gif = { git = "https://github.com/LedgerHQ/sdk_include_gif" }
```
Comment (xchapron-ledger) [inline Cargo.toml:12]: Would be great to migrate to the new crate which includes some unifications

File: Cargo.toml [hunk 2]
```diff
@@ -16,11 +17,11 @@ sbor = { path = "./sbor", default-features = false }
...
 
 [profile.dev]
-opt-level = "z"
+opt-level = "s"
 lto = false
 
 [profile.release]
-opt-level = "z"
+opt-level = "s"
 lto = false
```
Comment (xchapron-ledger) [inline Cargo.toml:25]: Maybe lto could be great to improve code size? Not sure what we stand on this @yhql ?

File: Cargo.toml (hunk 1 above)
Comment (xchapron-ledger) [reply Cargo.toml:12]: see https://github.com/LedgerHQ/app-boilerplate-rust/blob/main/Cargo.toml#L8-L10

**Review by xchapron-ledger — COMMENTED** [review]
Can you add Ledger reusable workflow? See https://github.com/LedgerHQ/app-boilerplate-rust/blob/main/.github/workflows/guidelines_enforcer.yml

File: Cargo.toml (hunk 2 above)
Comment (siy) [reply Cargo.toml:25]: "s" is compatible with the compilation options used for SDK

Comment (yhql) [reply Cargo.toml:25]: `lto` should help indeed, did you try enabling it @siy ?

File: Cargo.toml (hunk 1 above)
Comment (siy) [reply Cargo.toml:12]: Migration is not a big issue. Problem is that new SDK changed the name of the ELF symbol from `.ledger.api_level` to `ledger.api_level`. Now `cargo-ledger` can't extract API level information properly and flashing to Nano S Plus fails with famous `6512` error code.

File: Cargo.toml (hunk 2 above)
Comment (siy) [reply Cargo.toml:25]: Yes. With `lto` enabled, linking of final application fails with missing symbols:
```
  = note: rust-lld: error: undefined symbol: core::panicking::panic::h7a00e13465b87caa
          >>> referenced by delegate.rs:0 (src/int/specialized_div_rem/delegate.rs:0)
          >>>               compiler_builtins-6fd51d3edc4bf0ca.compiler_builtins.654993bf403450e3-cgu.3.rcgu.o:(compiler_builtins::int::specialized_div_rem::u64_div_rem::hbd91fc4fd5b90a92) in archive /home/siy/CLionProjects/Radix/Parallel/babylon-ledger-app/target/nanos/debug/deps/libcompiler_builtins-6fd51d3edc4bf0ca.rlib
          >>> referenced by binary_long.rs:0 (src/int/specialized_div_rem/binary_long.rs:0)
          >>>               compiler_builtins-6fd51d3edc4bf0ca.compiler_builtins.654993bf403450e3-cgu.3.rcgu.o:(compiler_builtins::int::specialized_div_rem::u32_div_rem::h4c4113bf8b1b018f) in archive /home/siy/CLionProjects/Radix/Parallel/babylon-ledger-app/target/nanos/debug/deps/libcompiler_builtins-6fd51d3edc4bf0ca.rlib
          >>> referenced by impls.rs:76 (src/mem/impls.rs:76)
          >>>               compiler_builtins-6fd51d3edc4bf0ca.compiler_builtins.654993bf403450e3-cgu.8.rcgu.o:(compiler_builtins::mem::memcpy::h641ca36c7c67fcf5) in archive /home/siy/CLionProjects/Radix/Parallel/babylon-ledger-app/target/nanos/debug/deps/libcompiler_builtins-6fd51d3edc4bf0ca.rlib
          >>> referenced 1 more times
```

File: Cargo.toml (hunk 1 above)
Comment (siy) [reply Cargo.toml:12]: Filed https://github.com/LedgerHQ/ledger-device-rust-sdk/issues/100

Comment (siy) [reply Cargo.toml:12]: And pushed a [PR](https://github.com/LedgerHQ/cargo-ledger/pull/34) for cargo-ledger

**Comment by siy** [comment]
> Can you add Ledger reusable workflow? See https://github.com/LedgerHQ/app-boilerplate-rust/blob/main/.github/workflows/guidelines_enforcer.yml

For some reason, it fails at metadata extraction phase. Any suggestions?

**Comment by xchapron-ledger** [comment]
> > Can you add Ledger reusable workflow? See https://github.com/LedgerHQ/app-boilerplate-rust/blob/main/.github/workflows/guidelines_enforcer.yml
> 
> For some reason, it fails at metadata extraction phase. Any suggestions?

Can you commit the workflow so that the checks run here and I can see if I can help you with the error.

**Comment by xchapron-ledger** [comment]
@siy I have fixed the workflow issue on our side, now you have errors that should be fixed on yours:
- build warning 
- missing definition in https://github.com/LedgerHQ/ledger-app-database/, regarding this, why are you requesting the 0x240 flags? Are they really needed? 0x200is needed for BLE on some devices, but doesn't need to be indicated here, 0x40 is only needed in specific case.

**Comment by xchapron-ledger** [comment]
Naive question here, do you really need to use features from ledger-sdk-secure-sys crate? This should be avoided. though I know as some palce the wrappers are missing in ledger-device-sdk crate.

**Comment by siy** [comment]
Added app to database https://github.com/LedgerHQ/ledger-app-database/pull/131

**Comment by siy** [comment]
> Naive question here, do you really need to use features from ledger-sdk-secure-sys crate? This should be avoided. though I know as some place the wrappers are missing in ledger-device-sdk crate.

The app contains a lot of functionality, way more than usually provided by average apps for other networks. This functionality requires a lot of RAM, which is a very scarce resource on Nano S. In many cases, I ought to use low level memory consumption optimization techniques which did not fit existing API's. So, reworking the application to use only "public" crates will make it useless because it will hang during transaction sign (ask me how I know :) ).

**Comment by siy** [comment]
@xchapron-ledger can you, please, trigger re-run of CI jobs? With the updated app DB they should pass.

**Comment by siy** [comment]
> Humf, the check is not really smart... can you update the curve order in this PR?

Fixed. All checks are green (except skipped one, obviously).

File: src/main.rs [hunk 3]
```diff
@@ -168,16 +167,20 @@ extern "C" fn sample_main() {
     let mut main_menu = Menu::new(&menu);
     let mut ticker = 0i8;
 
-    nanos_ui::ui::popup("Pending Review");
+    ledger_device_sdk::ui::gadgets::popup("Pending Review");
```
Comment (xchapron-ledger) [inline src/main.rs:170]: Would be better if using https://github.com/LedgerHQ/app-boilerplate-rust/blob/main/src/main.rs#L40

File: Cargo.toml [hunk 4]
```diff
@@ -1,31 +1,31 @@
...
 authors = ["siy"]
 edition = "2021"
 description = "Radix Babylon"
 
 [dependencies]
-nanos_sdk = { git = "https://github.com/LedgerHQ/ledger-nanos-sdk.git" }
-nanos_ui = { git = "https://github.com/radixdlt/ledger-nanos-ui-radix.git" }
-include_gif = { git = "https://github.com/LedgerHQ/sdk_include_gif" }
+ledger_device_sdk = "1.1.0"
+ledger_secure_sdk_sys = "1.0.2"
```
Comment (yogh333) [inline Cargo.toml:10]: Please use:
`ledger_device_sdk = "1.4.1"`
`ledger_secure_sdk_sys = "1.0.2"`

Comment (yogh333) [reply Cargo.toml:10]: Hello,

 FYI I did some tests after using latest Ledger Rust SDK and merging the PR into `develop` branch and it LGTM (performed somes tests on physical HW for both Nano S and S+). So please update the SDK versions so we can merge your PR ;)

File: src/main.rs (hunk 3 above)
Comment (siy) [reply src/main.rs:170]: Done

File: Cargo.toml (hunk 4 above)
Comment (siy) [reply Cargo.toml:10]: Done. Will push remaining changes after internal testing.

Comment (siy) [reply Cargo.toml]: Pushed remaining changes.

**Review by yogh333 — APPROVED** [review]
LGTM

**Pull request merged**